r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs

```diff
@@ -712,6 +720,158 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         Ok(result)
     }
 
+    /// Multiply `self` and `other` with an optimized algorithm that relies on the operands being
+    /// gadgets of a prime field with pseudo-mersenne modulus p, i.e., p = 2^n-c for a small
+    /// integer c. This optimization allows to obtain a result with limbs which are sufficiently
+    /// small to be conveniently represented without the intermediate representations,
+    /// i.e, as `NonNativeFieldGadget`.
+    /// To enable a further reduction, it is required that
+    /// ``
+    ///     bits_per_limb + surfeit' <= CAPACITY - 3,
+    /// ``
+    /// where
+    /// ``
+    ///     h = num_limbs*bits_per_limb - n
+    ///     surfeit' = log(num_adds(prod) + 3) =
+    ///             = log(num_limbs*(num_add(L)+1)*(num_add(R)+1)*(c*2^h+1)*2^bits_per_limb - 1 + 3)
+    ///             = log(num_limbs*(num_add(L)+1)*(num_add(R)+1)*(c*2^h+1)*2^bits_per_limb + 2)
+    /// ``
+    // cost: num_limbs^2 constraints
+    fn mul_without_prereduce_for_pseudomersenne<CS: ConstraintSystemAbstract<ConstraintF>>(
+        &self,
+        mut cs: CS,
+        other: &Self,
+        is_other_constant: bool, // Can be removed if a method `is_constant` is added to FpGadget, may be addressed in later refactoring
```

Comment:
I decided to employ a flag rather than creating a new function as most of the code is the same

Reply:
Well it kind of mixes the logic. I suggest to group as most common code as possible in the public function that then calls private implementations, as usual

Reply:
Not clear if mixing the logic you mean for the user of for the implementation. To clarify, consider that the flag is just used in private functions, there are still public functions `mul_by_constant`  and `mul_by_constant_without_prereduce` which then call the `mul_without_prereduce_for_pseudomersenne` with the flag being set to `true`. So this flag is never publicly exposed to the user.

Reply:
I understand what you mean, I just don't like this being different from what we have currently, for non mersenne fields, where there are to separate functions, mul_without_prereduce and mul_by_constant_without_prereduce.
So either you do the same for pseudo-mersenne (separating the logic) or you unify the two methods for non pseudo-mersenne (using the same boolean flag), you choose

Reply:
I added the flag also to the non pseudo-mersenne functions to unify the logic. I preferred this solution over having two distinct functions because it allows better maintainability in my opinion